README.md
<div align="center">

# Bevy Rewind 

A plugin for easily rewinding time in [Bevy](https://github.com/bevyengine/bevy)

[<img alt="bevy tracking" src="https://img.shields.io/badge/Bevy%20tracking-released%20version-lightblue?style=for-the-badge" height="24">](https://github.com/bevyengine/bevy/blob/main/docs/plugins_guidelines.md#main-branch-tracking)
[<img alt="build status" src="https://img.shields.io/github/actions/workflow/status/LiamGallagher737/bevy_rewind/rust.yml?branch=main&style=for-the-badge" height="24">](https://github.com/LiamGallagher737/bevy_rewind/actions)
[<img alt="github" src="https://img.shields.io/badge/github-bevy_rewind-8da0cb?style=for-the-badge&labelColor=555555&logo=github" height="24">](https://github.com/LiamGallagher737/bevy_rewind)
[<img alt="crates.io" src="https://img.shields.io/crates/v/bevy_rewind.svg?style=for-the-badge&color=fc8d62&logo=rust" height="24">](https://crates.io/crates/bevy_rewind)
[<img alt="docs.rs" src="https://img.shields.io/badge/docs.rs-bevy_rewind-66c2a5?style=for-the-badge&labelColor=555555&logo=docs.rs" height="24">](https://docs.rs/bevy_rewind)

</div>



# [Demo](examples/demo.rs)

Demo of rewinding using bevy_rapier, tap <kbd>Space</kbd> to toggle rewinding

```console
cargo run --example demo 
```

<div align="center">
    <img src="examples/images/demo.gif" alt="Gif of Demo Example" width="600" />
</div>



# Setup

Add the plugin to your app
```rust
.add_plugin(RewindPlugin::default())
```

Config Options

```rust
RewindPlugin {
    // How many captures will take place before they start clearing,
    // default is 300 for 5 seconds of replay.
    // (60 ticks per seconds * 5 seconds)
    max_capture_count: 300,
    // If the game should stop rewinding once any of the components has run out of history.
    cancel_on_empty_history: false, 
}
```



# Usage

To track a components value add the `RewindComponent<C>` to the entity with C being the component you want to track, in this example the entities `Transform` will be tracked for rewinding.

```rust
commands.spawn((
    PbrBundle::default(),
    RewindComponent::<Transform>::default(),
));
```

By default only `Transform` and `GlobalTransform` components can be tracked, to add support for more use the `init_rewind_component<C>` method on your app where C is the component you want to be able to track.

```rust
App::new()
    .add_plugin(RewindPlugin)
    .init_rewind_component::<Velocity>()
    .run();
```



# Bevy Tracking

|Bevy|bevy_rewind|
|---|---|
|0.10|0.1.0|



# License

All code in this repository is dual-licensed under either:

* MIT License ([LICENSE-MIT](LICENSE-MIT) or [http://opensource.org/licenses/MIT](http://opensource.org/licenses/MIT))
* Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or [http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0))

at your option. This means you can select the license you prefer! This dual-licensing approach is the de-facto standard in the Rust ecosystem and there are [very good reasons](https://github.com/bevyengine/bevy/issues/2373) to include both.

Unless you explicitly state otherwise, any contribution intentionally submitted for inclusion in the work by you, as defined in the Apache-2.0 license, shall be dual licensed as above, without any additional terms or conditions.


CHANGELOG.md
# Changelog

***Dates are formatted as DD.MM.YYYY***

<!-- Changelog Entry Template

## Version X.X.X (DD.MM.YYYY)

### Added

### Changed

### Removed

### Fixed

-->


examples/demo.rs
use bevy::prelude::*;
use bevy_rapier3d::prelude::*;
use bevy_rewind::*;

#[rustfmt::skip]
fn main() {
    println!();
    println!("Hold the Space key to toggle rewinding");
    println!();

    App::new()
        .add_plugins(DefaultPlugins)
        .add_plugin(RapierPhysicsPlugin::<NoUserData>::default())

        // Setup the scene
        .add_system(setup.on_startup())

        // Enable rewind when holding the space key
        .add_system(toggle_rewinding)

        // Enable rigidbodies when bot rewinding
        .add_system(enable_rigidbodies.run_if(not(rewinding)))

        // Disable rigidbodies when rewinding so rapier doesn't try to simulate physics when rewinding
        .add_system(disable_rigidbodies.run_if(rewinding))

        // Add the rewind plugin
        .add_plugin(RewindPlugin::new(300, true))

        // Add rewind surport for Velocity component
        .init_rewind_component::<Velocity>()

        .run();
}

fn setup(
    mut commands: Commands,
    mut meshes: ResMut<Assets<Mesh>>,
    mut materials: ResMut<Assets<StandardMaterial>>,
) {
    // Spawn Camera
    commands.spawn(Camera3dBundle {
        transform: Transform::from_translation(Vec3::new(20.0, 20.0, 20.0))
            .looking_at(Vec3::new(0.0, 0.0, 0.0), Vec3::Y),
        ..Default::default()
    });

    // Spawn Light
    commands.spawn(DirectionalLightBundle {
        transform: Transform::default().looking_at(Vec3::new(1.0, -3.0, 1.0), Vec3::Y),
        ..Default::default()
    });

    // Spawn Ground
    commands.spawn((
        PbrBundle {
            mesh: meshes.add(shape::Box::new(40.0, 0.5, 40.0).into()),
            material: materials.add(Color::WHITE.into()),
            transform: Transform::from_xyz(0.0, -0.5, 0.0),
            ..Default::default()
        },
        Collider::cuboid(20.0, 0.5, 20.0),
    ));

    // Spawn Spheres

    let mesh_handle = meshes.add(shape::UVSphere::default().into());
    let material_handle = materials.add(Color::RED.into());

    for n in 0..10 {
        commands.spawn((
            PbrBundle {
                mesh: mesh_handle.clone(),
                material: material_handle.clone(),
                transform: Transform::from_xyz(n as f32 * 0.1, n as f32 * 2.5, -n as f32 * 0.1),
                ..Default::default()
            },
            Collider::ball(1.0),
            RigidBody::Dynamic,
            Velocity::default(),
            RewindComponent::<Transform>::default(), // Track the Transform component
            RewindComponent::<Velocity>::default(),  // Track the Velocity component
        ));
    }
}

// Systems needed for the demo

fn toggle_rewinding(mut rewinding: ResMut<Rewinding>, keys: Res<Input<KeyCode>>) {
    if keys.just_pressed(KeyCode::Space) {
        **rewinding = !**rewinding;
    }
}

fn enable_rigidbodies(mut query: Query<&mut RigidBody>) {
    for mut rb in &mut query {
        *rb = RigidBody::Dynamic;
    }
}

fn disable_rigidbodies(mut query: Query<&mut RigidBody>) {
    for mut rb in &mut query {
        *rb = RigidBody::Fixed;
    }
}


src/lib.rs
#![warn(missing_docs)]

//! <div align="center">
//!
//! # Bevy Rewind
//!
//! A plugin for easily rewinding time in [Bevy](https://github.com/bevyengine/bevy)
//!
//! [<img alt="bevy tracking" src="https://img.shields.io/badge/Bevy%20tracking-released%20version-lightblue?style=for-the-badge" height="24">](https://github.com/bevyengine/bevy/blob/main/docs/plugins_guidelines.md#main-branch-tracking)
//! [<img alt="build status" src="https://img.shields.io/github/actions/workflow/status/LiamGallagher737/bevy_rewind/rust.yml?branch=main&style=for-the-badge" height="24">](https://github.com/LiamGallagher737/bevy_rewind/actions)
//! [<img alt="github" src="https://img.shields.io/badge/github-bevy_rewind-8da0cb?style=for-the-badge&labelColor=555555&logo=github" height="24">](https://github.com/LiamGallagher737/bevy_rewind)
//! [<img alt="crates.io" src="https://img.shields.io/crates/v/bevy_rewind.svg?style=for-the-badge&color=fc8d62&logo=rust" height="24">](https://crates.io/crates/bevy_rewind)
//! [<img alt="docs.rs" src="https://img.shields.io/badge/docs.rs-bevy_rewind-66c2a5?style=for-the-badge&labelColor=555555&logo=docs.rs" height="24">](https://docs.rs/bevy_rewind)
//!
//! </div>
//!
//! # Setup
//!
//! Add the plugin to your app
//! ```text
//! .add_plugin(RewindPlugin::default())
//! ```
//!
//! Config Options
//!
//! ```no_run
//! # use bevy_rewind::RewindSettings;
//! RewindSettings {
//!     // How many captures will take place before they start clearing,
//!     // default is 300 for 5 seconds of replay.
//!     // (60 ticks per seconds * 5 seconds)
//!     max_capture_count: 300,
//!     // If the game should stop rewinding once any of the components has run out of history.
//!     cancel_on_empty_history: false,
//! };
//! ```
//!
//! # Usage
//!
//! To track a components value add the `RewindComponent<C>` to the entity with C being the component you want to track, in this example the entities `Transform` will be tracked for rewinding.
//!
//! ```text
//! commands.spawn((
//!     PbrBundle::default(),
//!     RewindComponent::<Transform>::default(),
//! ));
//! ```
//!
//! By default on `Transform` and `GlobalTransform` can be tracked, to add mode use the `init_rewind_component<C>` method on your app where C is the component you want to be able to track.
//!
//! ```rust
//! # use bevy::prelude::*;
//! # use bevy_rewind::*;
//! # #[derive(Component, Clone)]
//! # struct Velocity;
//! App::new()
//!     .add_plugin(RewindPlugin::default())
//!     .init_rewind_component::<Velocity>()
//!     .run();
//! ```

use bevy::prelude::*;

/// Add this to your app for the ability to rewind
#[derive(Default)]
pub struct RewindPlugin {
    /// The [`RewindSettings`] to start the app with
    pub settings: RewindSettings,
}

impl RewindPlugin {
    /// Constructor for [`RewindPlugin`]
    pub fn new(max_capture_count: usize, cancel_on_empty_history: bool) -> Self {
        Self {
            settings: RewindSettings::new(max_capture_count, cancel_on_empty_history),
        }
    }
}

impl Plugin for RewindPlugin {
    fn build(&self, app: &mut App) {
        app.insert_resource(self.settings.clone());
        app.insert_resource(Rewinding::default());
        app.init_rewind_component::<Transform>();
        app.init_rewind_component::<GlobalTransform>();
    }
}

#[doc(hidden)]
pub trait RewindAppExtentions {
    /// Add rewindability for a given component
    fn init_rewind_component<C: Component + Clone>(&mut self) -> &mut Self;
}

impl RewindAppExtentions for App {
    fn init_rewind_component<C: Component + Clone>(&mut self) -> &mut Self {
        self.add_systems((
            capture_components::<C>
                .run_if(not(rewinding))
                .in_schedule(CoreSchedule::FixedUpdate),
            rewind_components::<C>
                .run_if(rewinding)
                .in_schedule(CoreSchedule::FixedUpdate),
        ));
        self
    }
}

/// Resource for controlling the rewind settings
#[derive(Resource, Clone)]
pub struct RewindSettings {
    /// The max number of captures per component before they start being cleared
    pub max_capture_count: usize,
    /// Stop rewinding if ***any*** [`RewindComponent`] has run out of history
    pub cancel_on_empty_history: bool,
}

impl RewindSettings {
    /// Constructor for [`RewindSettings`]
    pub fn new(max_capture_count: usize, cancel_on_empty_history: bool) -> Self {
        Self {
            max_capture_count,
            cancel_on_empty_history,
        }
    }
}

impl Default for RewindSettings {
    fn default() -> Self {
        Self {
            max_capture_count: 300,
            cancel_on_empty_history: false,
        }
    }
}

/// Resource for controlling rewinding
#[derive(Resource, Default, Deref, DerefMut)]
pub struct Rewinding(bool);

/// Add this component to any entity to track it for rewinding, the generic value is the component you want to track
#[derive(Component, Default)]
pub struct RewindComponent<C: Component + Clone> {
    history: Vec<C>,
}

/// Run condition that returns true if rewinding is true
pub fn rewinding(rewinding: Res<Rewinding>) -> bool {
    **rewinding
}

fn capture_components<C: Component + Clone>(
    mut query: Query<(&C, &mut RewindComponent<C>)>,
    rewind_settings: Res<RewindSettings>,
) {
    for (component, mut rewind) in &mut query {
        rewind.history.push(component.clone());
        if rewind.history.len() > rewind_settings.max_capture_count {
            rewind.history.remove(0);
        }
    }
}

fn rewind_components<C: Component + Clone>(
    mut query: Query<(&mut C, &mut RewindComponent<C>)>,
    mut rewinding: ResMut<Rewinding>,
    rewind_settings: Res<RewindSettings>,
) {
    for (mut component, mut rewind) in &mut query {
        if let Some(old_value) = rewind.history.pop() {
            *component = old_value;
        } else if rewind_settings.cancel_on_empty_history {
            **rewinding = false;
        }
    }
}
